crates/wasmtime/src/memory.rs

```diff
@@ -654,6 +714,116 @@ pub unsafe trait MemoryCreator: Send + Sync {
+        if !ty.is_shared() {
+            bail!("shared memory must have the `shared` flag enabled on its memory type")
+        }
+        debug_assert!(ty.maximum().is_some());
+
+        let tunables = &engine.config().tunables;
+        // TODO: check the tunables (only static allocation allowed)
+        let memory = wasmtime_runtime::SharedMemory::new(ty.wasmtime_memory().clone(), tunables)?;
+        Ok(Self(memory, engine.clone()))
+    }
+
+    /// Return a reference to the [`Engine`] used to configure the shared
+    /// memory.
+    pub fn engine(&self) -> &Engine {
+        &self.1
+    }
+
+    /// Return the type of the shared memory.
+    pub fn ty(&self) -> MemoryType {
+        MemoryType::from_wasmtime_memory(&self.0.ty())
+    }
+
+    /// Returns the size, in WebAssembly pages, of this wasm memory.
+    pub fn size(&self) -> u64 {
+        (self.data_size() / wasmtime_environ::WASM_PAGE_SIZE as usize) as u64
+    }
+
+    /// Returns the byte length of this memory.
+    ///
+    /// The returned value will be a multiple of the wasm page size, 64k.
+    ///
+    /// For more information and examples see the documentation on the
+    /// [`Memory`] type.
+    pub fn data_size(&self) -> usize {
+        use wasmtime_runtime::RuntimeLinearMemory;
+        self.0.byte_size()
+    }
+
+    /// Return read access to the available portion of the shared memory. Note
+    /// that the available pages may be between `[minimum, maximum]`.
+    pub fn data(&self) -> &[u8] {
+        unsafe {
+            let definition = *self.0.vmmemory_ptr();
+            slice::from_raw_parts(definition.base, definition.current_length)
+        }
+    }
+
+    /// Return write access to the available portion of the shared memory. Note
+    /// that the available pages may be between `[minimum, maximum]`.
+    pub fn data_mut(&mut self) -> &mut [u8] {
+        unsafe {
+            let definition = *self.0.vmmemory_ptr_mut();
+            slice::from_raw_parts_mut(definition.base, definition.current_length)
+        }
+    }
+
+    /// Grows this WebAssembly memory by `delta` pages.
+    ///
+    /// This will attempt to add `delta` more pages of memory on to the end of
+    /// this `Memory` instance. If successful this may relocate the memory and
+    /// cause [`Memory::data_ptr`] to return a new value. Additionally any
+    /// unsafely constructed slices into this memory may no longer be valid.
+    ///
+    /// On success returns the number of pages this memory previously had
+    /// before the growth succeeded.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if memory could not be grown, for example if it exceeds
+    /// the maximum limits of this memory. A
+    /// [`ResourceLimiter`](crate::ResourceLimiter) is another example of
+    /// preventing a memory to grow.
+    pub fn grow(&self, delta: u64) -> Result<u64> {
+        todo!()
+        // self.0.grow(delta, <need to provide a store>)
```

Comment:
I think a refactoring to make this work would be to:
 - change the signature of `RuntimeLinearMemory::grow` to take a `limiter_fn: Option<Fn<...>>` instead of the entire store
 - remove the error reporting to the limiter (`Store::memory_grow_failed` -> `ResourceLimiter::memory_grow_failed`); right now it is unclear (except during testing) why a limiter would need to receive the growth failure--shouldn't it be the one producing this error instead? If indeed this _is_ needed, perhaps we could alter the suggestion above to an `Option<dyn ResourceLimiter>`?

Reply:
Oh sorry I sort of independently came to a similar conclusion that this would be a problem above before I saw your comment here.

I think for now let's skip limiting memory growth of shared memories with `ResourceLimiter` so this would thread through a `None` or similar (and wherever uses the limiter would assert the store is present if the memory isn't shared).

Overall we're going to need a different strategy for limiting shared memories since growth would notify a random store which isn't necessarily the right "store". I would naively say that we'll need a limiter-per-shared-linear-memory but we can figure that out later, no need to have it covered for now.